core/message.go

```diff
@@ -185,72 +202,92 @@ func (msg *Message) Clone() *Message {
 func (msg *Message) CloneOriginal() *Message {
-	clone := *msg
+	if msg.orig == nil {
+		msg.FreezeOriginal()
+	}
 
+	clone := *msg
 	clone.data.payload = MessageDataPool.Get(len(msg.orig.payload))
 	copy(clone.data.payload, msg.orig.payload)
 
-	clone.SetStreamID(msg.orig.streamID)
+	if msg.orig.Metadata == nil {
+		clone.data.Metadata = msg.orig.Metadata.Clone()
+	} else {
+		clone.data.Metadata = nil
+	}
 
+	clone.SetStreamID(msg.origStreamID)
 	return &clone
 }
 
-// FreezeOriginal set the original data and freeze the message
+// FreezeOriginal will take the current state of the message and store it as
+// the "original" message. This function can only be called once for each
+// message. Please note that this function only affects payload and metadata and
+// can only be called once. Additional calls will have no effect.
+// The original stream ID can be changed at any time by using SetOrigStreamID.
 func (msg *Message) FreezeOriginal() {
-	msg.orig.payload = getPayloadCopy(msg.data.payload)
-	msg.orig.streamID = msg.data.streamID
+	// Freeze may be called only once
+	if msg.orig != nil {
+		return
+	}
 
-	if msg.data.Metadata == nil {
-		msg.orig.Metadata = nil
-	} else {
-		msg.orig.Metadata = msg.data.Metadata.Clone()
+	var metadata Metadata
+	if msg.data.Metadata != nil {
+		metadata = msg.data.Metadata.Clone()
+	}
+
+	msg.orig = &MessageData{
+		payload:  getPayloadCopy(msg.data.payload),
+		Metadata: metadata,
 	}
 }
 
 // Serialize generates a new payload containing all data that can be preserved
 // over shutdown (i.e. no data directly referencing runtime components). The
-// serialized data is based on the current message state.
+// serialized data is based on the current message state and does not perserve
+// the original data created by FreezeOriginal.
 func (msg *Message) Serialize() ([]byte, error) {
-	return msg.data.serialize(msg.prevStreamID, msg.timestamp)
+	return msg.data.serialize(msg)
 }
 
 // SerializeOriginal generates a new payload containing all data that can be
-// preserved over shutdown (i.e. no data directly referencing runtime c
-// omponents). The serialized data is based on the original message
+// preserved over shutdown (i.e. no data directly referencing runtime
+// components). The serialized data is based on the original message.
 func (msg *Message) SerializeOriginal() ([]byte, error) {
-	return msg.orig.serialize(msg.data.streamID, msg.timestamp)
+	return msg.orig.serialize(msg)
 }
 
-func (data MessageData) serialize(prevStreamID MessageStreamID, timestamp time.Time) ([]byte, error) {
+func (data MessageData) serialize(msg *Message) ([]byte, error) {
 	serializable := &SerializedMessage{
-		StreamID:     proto.Uint64(uint64(data.streamID)),
-		PrevStreamID: proto.Uint64(uint64(prevStreamID)),
-		Timestamp:    proto.Int64(timestamp.UnixNano()),
+		StreamID:     proto.Uint64(uint64(msg.streamID)),
+		PrevStreamID: proto.Uint64(uint64(msg.prevStreamID)),
```

Comment:
Not there as it wasn't there before.
This is a tricky one.
We did not add the original data to serialization when we introduced it as this would have broken deserialization of data written with 0.4.x. In that case we would get data without original data but act as if there was. False assumption = wrong behavior = broken pipelines.

Now looking at that decision in retrostpective I think we should change this.
The pipelines will change drastically anyway, so you should see that you drain any serialized data before switching to 0.5.0 (this should be added to the release notes).
But this should be a PR on its own.